### PR TITLE
Add tests

### LINKED ISSUE #4: Add test units for alert service

### PR DESCRIPTION
**Pull request by cirelli94**
Fix #4

### CODE REVIEW

File: fixture/test_dataservice_mongohostdata_03.json [hunk 1]
```diff
@@ -0,0 +1,432 @@
+{
+        "ID": "5dd41aa2b2fa40163c878538",
```
Comment (amreo) [inline fixture/test_dataservice_mongohostdata_03.json]: It has the same `ID` and `CreatedAt` as `fixture/test_dataservice_mongohostdata_02.json`

File: fixture/test_dataservice_mongohostdata_02.json [hunk 2]
```diff
@@ -0,0 +1,432 @@
+{
+        "ID": "5dd41aa2b2fa40163c878538",
```
Comment (amreo) [inline fixture/test_dataservice_mongohostdata_02.json]: It has the same id as `fixture/test_dataservice_mongohostdata_02.json`

File: alert-service/service/service_test.go [hunk 3]
```diff
@@ -28,6 +30,94 @@ import (
...
+			AlertService: config.AlertService{LogMessages: true}},
+		Log: utils.NewLogger("TEST"),
+	}
+
+	msg := hub.Message{
+		Name: "",
+	}
+
+	as.ProcessMsg(msg)
+}
```
Comment (amreo) [inline alert-service/service/service_test.go]: What are you verifying?

File: alert-service/service/service_test.go [hunk 4]
```diff
@@ -114,6 +204,128 @@ func TestProcessHostDataInsertion_DiffHostError3(t *testing.T) {
...
+	params["alert"] = model.Alert{
+		OtherInfo:     map[string]interface{}{},
+		AlertSeverity: model.AlertSeverityMajor,
+		Description:   "This is just an alert test to a mocked emailer.",
+		Date:          utils.P("2019-09-02T10:25:28Z"),
+		AlertCode:     model.AlertCodeNewLicense,
+	}
+
+	as.ProcessAlertInsertion(params)
+}
```
Comment (amreo) [inline alert-service/service/service_test.go]: What are you verifying?

**Review by amreo — COMMENTED** [review]
Can you a add another utils.LoadFixtureHostdata for MongoHostdata?

File: alert-service/database/database_test.go [hunk 5]
```diff
@@ -38,6 +43,100 @@ func TestMongodbSuite(t *testing.T) {
...
+		Log: logger,
+	}
+
+	assert.PanicsWithValue(t, "log.Fatal called by test", db.ConnectToMongodb)
+}
+
+func (m *MongodbSuite) TestFindHostData_SuccessExist() {
+	defer m.db.Client.Database(m.dbname).Collection("hosts").DeleteMany(context.TODO(), bson.M{})
+
+	hd := utils.LoadFixtureHostData(m.T(), "../../fixture/test_dataservice_mongohostdata_01.json")
```
Comment (amreo) [inline alert-service/database/database_test.go]: Can you use a "utils.LoadFixtureMongoHostdata"?

File: alert-service/database/database_test.go [hunk 6]
```diff
@@ -38,6 +43,100 @@ func TestMongodbSuite(t *testing.T) {
...
+	err := m.InsertHostData(hd)
```
Comment (amreo) [inline alert-service/database/database_test.go]: Can you fix InsertHostData to accept a map[string]interface{} param?

File: alert-service/database/database_test.go [hunk 7]
```diff
@@ -38,6 +43,100 @@ func TestMongodbSuite(t *testing.T) {
...
+}
+
+func (m *MongodbSuite) TestFindHostData_FailWrongID() {
+	defer m.db.Client.Database(m.dbname).Collection("hosts").DeleteMany(context.TODO(), bson.M{})
+
+	hd := utils.LoadFixtureHostData(m.T(), "../../fixture/test_dataservice_mongohostdata_01.json")
+	err := m.InsertHostData(hd)
+	require.NoError(m.T(), err)
+
+	notExistingID := primitive.NewObjectIDFromTimestamp(time.Now())
```
Comment (amreo) [inline alert-service/database/database_test.go]: Can you make the test more reproducibile providing a fixed ObjectID? In utils/test.go exist a function called `Str2oid`

File: alert-service/database/database_test.go [hunk 8]
```diff
@@ -38,6 +43,100 @@ func TestMongodbSuite(t *testing.T) {
...
+
+	assert.Equal(m.T(), model.HostData{}, hd2)
+}
+
+func (m *MongodbSuite) TestFindMostRecentHostDataOlderThan_OneInsert_Success() {
+	defer m.db.Client.Database(m.dbname).Collection("hosts").DeleteMany(context.TODO(), bson.M{})
+
+	hd := utils.LoadFixtureHostData(m.T(), "../../fixture/test_dataservice_hostdata_01.json")
+	hd.CreatedAt = time.Now().UTC().Truncate(time.Millisecond)
+	err := m.InsertHostData(hd)
```
Comment (amreo) [inline alert-service/database/database_test.go]: Can you fix the CreatedAt at fixture level?

File: alert-service/database/database_test.go [hunk 9]
```diff
@@ -38,6 +43,100 @@ func TestMongodbSuite(t *testing.T) {
...
+	require.NoError(m.T(), err)
+
+	time.Sleep(1 * time.Second)
```
Comment (amreo) [inline alert-service/database/database_test.go]: NO! you can make a temporal warp providing a custom timeNow to the testing m.db.timeNow

File: alert-service/database/database_test.go [hunk 10]
```diff
@@ -38,6 +43,100 @@ func TestMongodbSuite(t *testing.T) {
...
+	afterFirstInsert := time.Now()
```
Comment (amreo) [inline alert-service/database/database_test.go]: Can you make the test more reproducibile using a fixed time.Time?

File: alert-service/database/database_test.go [hunk 11]
```diff
@@ -38,6 +43,100 @@ func TestMongodbSuite(t *testing.T) {
...
+
+	hd := utils.LoadFixtureHostData(m.T(), "../../fixture/test_dataservice_mongohostdata_01.json")
+	hd.ID = primitive.NewObjectIDFromTimestamp(time.Now())
+	hd.CreatedAt = utils.P("2019-11-05T14:02:03Z")
+	err := m.InsertHostData(hd)
+	require.NoError(m.T(), err)
+
+	m.ArchiveHost(hd.Hostname)
+	hd.Archived = true
+
```
Comment (amreo) [inline alert-service/database/database_test.go]: Can you archive the mongohostdata at fixture level?

File: alert-service/database/database_test.go [hunk 12]
```diff
@@ -38,6 +43,100 @@ func TestMongodbSuite(t *testing.T) {
...
+	require.NoError(m.T(), err2)
+	assert.Equal(m.T(), hd, hd2)
+}
+
+func (m *MongodbSuite) TestFindMostRecentHostDataOlderThan_MoreInserts_Success() {
+	defer m.db.Client.Database(m.dbname).Collection("hosts").DeleteMany(context.TODO(), bson.M{})
+
+	hd := utils.LoadFixtureHostData(m.T(), "../../fixture/test_dataservice_mongohostdata_01.json")
+	hd.ID = primitive.NewObjectIDFromTimestamp(time.Now())
+	hd.CreatedAt = utils.P("2019-11-05T14:02:03Z")
```
Comment (amreo) [inline alert-service/database/database_test.go]: Can you fix the CreatedAt at fixture level?

File: alert-service/database/database_test.go [hunk 13]
```diff
@@ -38,6 +43,100 @@ func TestMongodbSuite(t *testing.T) {
...
+	hd.Archived = true
+
+	afterFirstInsert := utils.P("2019-11-05T15:02:03Z")
+	hd3, err3 := m.db.FindMostRecentHostDataOlderThan("itl-csllab-112.sorint.localpippo", afterFirstInsert)
+	require.NoError(m.T(), err3)
+	assert.Equal(m.T(), hd, hd3)
+
+	hd2 := utils.LoadFixtureHostData(m.T(), "../../fixture/test_dataservice_mongohostdata_02.json")
+	hd2.ID = primitive.NewObjectIDFromTimestamp(time.Now())
+	hd2.CreatedAt = utils.P("2019-11-06T14:02:03Z")
```
Comment (amreo) [inline alert-service/database/database_test.go]: Idem

File: alert-service/database/database_test.go [hunk 14]
```diff
@@ -51,7 +150,46 @@ func (m *MongodbSuite) TestInsertAlert_Success() {
...
 
 	assert.Equal(m.T(), alert1, out)
+}
+
+func (m *MongodbSuite) TestFindOldCurrentHosts() {
+	defer m.db.Client.Database(m.dbname).Collection("hosts").DeleteMany(context.TODO(), bson.M{})
+
+	hd := utils.LoadFixtureHostData(m.T(), "../../fixture/test_dataservice_mongohostdata_01.json")
+	hd.ID = primitive.NewObjectIDFromTimestamp(time.Now())
+	hd.CreatedAt = utils.P("2019-11-05T14:02:03Z")
```
Comment (amreo) [inline alert-service/database/database_test.go]: Idem

File: alert-service/database/database_test.go [hunk 15]
```diff
@@ -51,7 +150,46 @@ func (m *MongodbSuite) TestInsertAlert_Success() {
...
+	err := m.InsertHostData(hd)
+	require.NoError(m.T(), err)
 
+	hd2 := utils.LoadFixtureHostData(m.T(), "../../fixture/test_dataservice_mongohostdata_03.json")
+	hd2.ID = primitive.NewObjectIDFromTimestamp(time.Now())
+	hd2.CreatedAt = utils.P("2020-01-05T14:02:03Z")
```
Comment (amreo) [inline alert-service/database/database_test.go]: Idem

File: alert-service/database/suite_test.go [hunk 16]
```diff
@@ -76,3 +77,16 @@ func (db *MongodbSuite) InsertHostData(hostData model.HostData) error {
...
+		"Hostname": hostname,
+		"Archived": false,
+	}, mu.UOSet(bson.M{
+		"Archived": true,
+	})); err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (amreo) [inline alert-service/database/suite_test.go]: This function is no longer needed if you fix the errors above

**Review by amreo — CHANGES_REQUESTED** [review]

**Comment by amreo** [comment]
Can you try to test the function HostDataInsertion in service.go?

File: fixture/test_dataservice_mongohostdata_02.json (hunk 2 above)
Comment (cirelli94) [reply fixture/test_dataservice_mongohostdata_02.json]: This IS fixture/test_dataservice_mongohostdata_02.json !

Comment (cirelli94) [reply fixture/test_dataservice_mongohostdata_02.json]: (I understood)

File: alert-service/service/service_test.go (hunk 3 above)
Comment (cirelli94) [reply alert-service/service/service_test.go]: That it does nothing: alertService is empty and would crash if ProcessMsg use something it shouldn't.
as.ProcessMsg(msg) shall only log a warn.

File: alert-service/service/service_test.go (hunk 4 above)
Comment (cirelli94) [reply alert-service/service/service_test.go]: I'm verifying that even if the emailer behave wrongly and return an error, this error is managed.

File: alert-service/database/database_test.go (hunk 5 above)
Comment (cirelli94) [reply alert-service/database/database_test.go]: What should be the differences with LoadFixtureHostData?

Comment (cirelli94) [reply alert-service/database/database_test.go]: Loading a map[string]interface{} instead of a model.HostData?

Comment (amreo) [reply alert-service/database/database_test.go]: Yes

**Review by amreo — APPROVED** [review]

**Pull request merged**